### ISSUE TITLE
Mini roundabouts

### ISSUE THREAD

**Issue by aubergine10** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#73]
This is where a junction of a _small_ road (specifically: 2-lane 2-way) is treated as a roundabout.

Usually there will be a white circle painted in the middle, and applicable street signs, to convey this to drivers.

![download](https://user-images.githubusercontent.com/1386719/52712743-67360b00-2f8d-11e9-96fc-c5dd11157b7e.jpg)

Drivers must drive around the circle, in the normal direction of traffic as depicted by the arrows painted on the road (which match the arrows on the road sign).

This greatly reduces the number of conflict points (see #4) at the junction, and helps improve traffic flow.

There is an additional configuration in cases where 4 or more roads need to be connected; two mini roundabout junctions can be used side by side (this is all about conflict point reduction):

![download](https://user-images.githubusercontent.com/1386719/52713050-3e624580-2f8e-11e9-8cc5-4c87c0aabfaa.jpg)

At some point in the future, I think ability to turn small junctions in to mini roundabouts would be a nice feature to implement.

**Comment by Strdate** [comment]
You can do it already with this mod ;) https://steamcommunity.com/sharedfiles/filedetails/?id=1625704117
Set the radius to five units and it will work similarly. Unfortunately the roads appear buggy.

**Comment by aubergine10** [comment]
> You can do it already with this mod ;) https://steamcommunity.com/sharedfiles/filedetails/?id=1625704117

No, that's very different, as it is conventional roundabout builder.

Specifically: A mini-roundabout exists *in the junction*. In other words, the lane routing within the junction would be circular (like a virtual circular lane with a node near each exit).

There are some decal assets to do the road marking for this, but currently all cars just drive over them because they see it as normal junction.

**Comment by Strdate** [comment]
If you build it really small then it is of a size of standard intersection but there is a roundabout inside (use highway ramp roads, they are small enuf). It looks hideous, but it is somehow working.

**Comment by aubergine10** [comment]
The decal I use is in this pack: https://steamcommunity.com/sharedfiles/filedetails/?id=1611818494

**Comment by Strdate** [comment]
@aubergine10 There it is: https://imgur.com/a/Hf6FBim

**Comment by kianzarrin** [comment]
@aubergine10 should we close this issue?
If not please explain what is lacking by roundabout builder? 
You did say you want to turn the junction intro roundabout but using roundabout with very small radius in roundabout builder has the same functional/visual influence.

**Comment by aubergine10** [comment]
No, this is different thing entirely. Specifically, it's _changing the paths vehicles take through the junction node_.

While roundabout builder with small radius might come close, it's not the same thing. Roundabout builder would likely have at least 2 or maybe more segments to represent the roundabout.

What this issue describes is changing the vehicle AI (I assume that's best place to do it) so that on a 'roundabout' node they'll drive round in a circular motion until they reach their exit, rather than cutting across other traffic. It's specifically for junctions of small roads (3-4 roads, 2-lane 2-way usually, but some 1-lane 1-way roads could be present too).

![image](https://user-images.githubusercontent.com/1386719/79683557-e0c92900-8222-11ea-9dcb-7031386e5154.png)

![image](https://user-images.githubusercontent.com/1386719/79683462-35b86f80-8222-11ea-83fc-ea1a287bbb9a.png)

![image](https://user-images.githubusercontent.com/1386719/79683470-479a1280-8222-11ea-8b0a-224a1c5bc214.png)

![image](https://user-images.githubusercontent.com/1386719/79683479-52ed3e00-8222-11ea-9505-af9ce34c7cc0.png)

The key benefit is that it takes almost zero extra space, because the roundabout is _in the node_.

It would mean any junction with 3-4 small roads could be turned in to a roundabout without altering the road layout.

It could be further embelished with your road blender mod, for example to widen the road as it enters the roundabout.

**Comment by kianzarrin** [comment]
@aubergine10 if you looked at @Strdate image the roundabout was not bigger than the junction so I still don't see what is the advantage of doing it your way.
BTW I am surprised segments so small in @Strdate image cause no problem!

**Comment by aubergine10** [comment]
Segments that small did cause issue hence the paved decal placed over the road.

The vehicle AI already has to handle path a vehicle takes across a node; the idea for mini roundabout is that it's just an alternate bit of code for sending vehicle across the node.

In other words, user doesn't need to alter their road network at all - they just set a flag "this junction is a mini roundabout". And to remove the mini roundabout they'd unset the flag.

There are potentially several "junction modes" we could consider:

- Vanilla - do what you want, try not to crash
- Mini roundabout (small roads only; although possible with medium road if outer lanes are limited to connecting to adjacent road only)
- Prioritised (priority signs)
- Box junction - https://github.com/CitiesSkylinesMods/TMPE/issues/424#issuecomment-553247660 (medium/large roads)
- Turn filters - #46 (medium/large roads)
- Traffic lights

So the user could try all those things (or combinations in most cases) without "physical" alterations to their roads. For me that's one of the best things about TMPE - I can change how traffic works, in most cases, without needing to alter roads or at least be able to minimise alterations to roads (eg. to add an extra lane to facilitate turning lane at junction).

**Comment by aubergine10** [comment]
BTW, I'm not seeing small roundabouts as a replacement for roundabout builder. When I want a larger roundabout, I'll head straight for roundabout builder.

The mini roundabouts are more of a "let's try alternate flow of traffic through this small road junction first, before I start altering my road network".

**Comment by aubergine10** [comment]
Another reason for not making mini roundabouts out of tiny road segments is that those segments/lanes/nodes all count towards game limit. If we could have the vehicle AIs treat a normal node as mini roundabout, that does not consume net limits.

**Comment by kianzarrin** [comment]
> @thebugfixnet said in https://github.com/kianzarrin/RoadTransitionManager/issues/3

https://github.com/kianzarrin/RoadTransitionManager/issues/3

While testing the Road Transition Manager (and i must say that i already love this mod!) the normal (and correct) behaviour is that the vehicles drive straight over this junctions:

![image](https://user-images.githubusercontent.com/37211016/81298363-f8067400-9074-11ea-89f6-700ed045b2ba.png)

But at a very large - very smoothed - node it would be great to have an option to manipulate the lanes in a way that they act somehowe like a road with an middle island (not really like a roundabout, as there should be no full turns possible). 

Maybe to have it easier w/o left turns:
![image](https://user-images.githubusercontent.com/37211016/81299478-92b38280-9076-11ea-82d9-68bce4db6fba.png)

Option 1 with left turns:
![image](https://user-images.githubusercontent.com/37211016/81299777-06558f80-9077-11ea-88de-b923ca765946.png)

Option 2 with left turns:
![image](https://user-images.githubusercontent.com/37211016/81299996-63e9dc00-9077-11ea-8aa0-0d83a0db20d8.png)

Why should be such an option a nice to have?

Just because of possibilities like this:
![image](https://user-images.githubusercontent.com/37211016/81301720-b3c9a280-9079-11ea-93b0-518a66e2214e.png)

or:
![image](https://user-images.githubusercontent.com/37211016/81302716-03f53480-907b-11ea-82ff-3d2db5c8bc02.png)

😄 

I don't know at all if lanes could be manipulated / bend in this way - but want just to share my thoughts :)

And maybe... some dreams come true....

**Comment by kianzarrin** [comment]
> And maybe... some dreams come true....

It certainly brings @aubergine10 's dream true. while I like the idea I think we can spend the time to do more important stuff. Particularly when you can achieve similar results by using roundabout builder https://imgur.com/a/Hf6FBim . just set the radius to very small and then paint over it like the picture.

**Comment by thebugfixnet** [comment]
>Particularly when you can achieve similar results by using roundabout builder https://imgur.com/a/Hf6FBim . just set the radius to very small and then paint over it like the picture.
>
This is a way to do - but at the moment with such tiny roundabounts i would not able use the road transition manager to smoothen the nodes - at least as long as it struggles with such short distances i showed you the last days...
 
> EDIT: I am mesmerized by your roundabouts!
Thanks :)

**Comment by aubergine10** [comment]
@kianzarrin These are exactly the sort of stuff I was thinking of - increased corner radius, and traffic flowing through the node as if there is an obstacle in middle of node:

<details><summary>images from @thebugfixnet </summary>

![ex1](https://user-images.githubusercontent.com/37211016/81301720-b3c9a280-9079-11ea-93b0-518a66e2214e.png)

![ex2](https://user-images.githubusercontent.com/37211016/81302716-03f53480-907b-11ea-82ff-3d2db5c8bc02.png)
</details>
